src/Microsoft.IdentityModel.Logging/LogHelper.cs

```diff
@@ -371,5 +418,47 @@ public static object MarkAsNonPII(object arg)
         {
             return new NonPII(arg);
         }
+
+        /// <summary>
+        /// Creates a <see cref="LogEntry"/> by using the provided event level, exception argument, string argument and arguments list.
+        /// </summary>
+        /// <param name="level"><see cref="EventLevel"/></param>
+        /// <param name="innerException"><see cref="Exception"/></param>
+        /// <param name="message">The log message.</param>
+        /// <param name="args">An object array that contains zero or more objects to format.</param>
+        private static LogEntry WriteEntry(EventLevel level, Exception innerException, string message, params object[] args)
+        {
+            if (innerException != null)
+            {
+                // if PII is turned off and 'innerException' is a System exception only display the exception type
+                if (!IdentityModelEventSource.ShowPII && !LogHelper.IsCustomException(innerException))
+                    message = string.Format(CultureInfo.InvariantCulture, "Message: {0}, InnerException: {1}", message, innerException.GetType());
+                else // otherwise it's safe to display the entire exception message
+                    message = string.Format(CultureInfo.InvariantCulture, "Message: {0}, InnerException: {1}", message, innerException.Message);
+            }
+
+            LogEntry entry = new LogEntry();
+
+            // Logs basic information (library version, DateTime, weather PII is ON/OFF) once before any log messages are written.
+            if (!HeaderWritten)
+            {
+                entry.EventLevel = EventLevel.LogAlways;
+                entry.Message = string.Format(CultureInfo.InvariantCulture, _versionLogMessage, typeof(IdentityModelEventSource).GetTypeInfo().Assembly.GetName().Version.ToString());
+                Logger.Log(entry);
+
+                entry.Message = string.Format(CultureInfo.InvariantCulture, _dateLogMessage, DateTime.UtcNow);
+                Logger.Log(entry);
+
+                entry.Message = IdentityModelEventSource.ShowPII ? _piiOnLogMessage : _piiOffLogMessage;
+                Logger.Log(entry);
+
+                HeaderWritten = true;
+            }
+
+            entry.EventLevel = level;
+            entry.Message = args != null ? FormatInvariant(message, args) : message;
```

Comment:
Will the message always be logged?
We are formatting the string.

Reply:
If message == null and args == null 
    entry.Message will be null
if message == null and args != null
    entry.Message will be string,Empty (as returned by FormatInvariant(..))

I will add additional checks here and improve LoggerAdapter to take this into consideration.

Reply:
Updated. Please review the method one more time.